### PR TITLE
fix(appRoles): don't allowed to upload duplicate app roles

### PR DESCRIPTION
**Pull request by tfjanjua**
## Description

- eliminated duplicate app role values from the APIs: /api/apps/appreleaseprocess/{appId}/role **&** /api/apps/AppChange/{appId}/role/activeapp

## Why

An app should have unique roles, duplicate roles are causing issue while assigning app roles to any company user.

## Issue

[Ref: #868](https://github.com/eclipse-tractusx/portal-backend/issues/868)

## Checklist

Please delete options that are not relevant.

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/docs/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Review by evegufy — DISMISSED** [review]
Hi @tfjanjua thank you for the contribution! could you please resolve the remaining codeQL findings "Useless assignment to local variable"?

File: src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs [hunk 1]
```diff
@@ -37,15 +37,34 @@
...
+        var appRoles = _dbContext.UserRoles.Where(w => appIdRoles.Select(a => a.AppId).Contains(w.OfferId))
+                                           .Select(x => new { AppId = x.OfferId, Role = x.UserRoleText })
+                                           .ToList();
+        List<UserRole>? uniqueRoles;
+        uniqueRoles = appRoles != null && appRoles.Any()
+        ?   // Selecting the roles, except those that are already part of an APP
+            uniqueRoles = appIdRoles.Select(x => new { x.AppId, x.Role })
+                        .Except(appRoles)
+                        .Select(x => new UserRole(Guid.NewGuid(), x.Role, x.AppId))
+                        .ToList()
```
[the remark this answers is not in the document]
Comment (tfjanjua) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs:49]: value of uniqueRoles variable has been used or read at line# 55 and 57

File: src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs [hunk 2]
```diff
@@ -37,15 +37,34 @@
...
+        :   // No roles has been attached to the APP so, all given roles would be inserting into the database
+            uniqueRoles = appIdRoles.Select(x => new UserRole(Guid.NewGuid(), x.Role, x.AppId)).ToList();
```
[the remark this answers is not in the document]
Comment (tfjanjua) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs:51]: value of uniqueRoles variable has been used or read at line# 55 and 57

**Comment by tfjanjua** [comment]
> Hi @tfjanjua thank you for the contribution! could you please resolve the remaining codeQL findings "Useless assignment to local variable"?

Hi @evegufy , its resolved

**Comment by evegufy** [comment]
Hi @ntruchsess @Phil91 could you please review?

File: src/marketplace/Apps.Service/Extensions/AppExtensions.cs [hunk 3]
```diff
@@ -58,11 +58,16 @@ public static void ValidateAppUserRole(Guid appId, IEnumerable<AppUserRole> appU
     /// <returns>returns the created appRoleData</returns>
     public static IEnumerable<AppRoleData> CreateUserRolesWithDescriptions(IUserRolesRepository userRolesRepository, Guid appId, IEnumerable<AppUserRole> userRoles) =>
         userRoles.Zip(
-            userRolesRepository.CreateAppUserRoles(userRoles.Select(x => (appId, x.Role))),
+            // Distinct() will eleminate the role with same Role Name value
+            // It will ba a bit confusing if Role Names are duplicate but description of role is different
+            // Description with multiple language code of a Role make sense
+            userRolesRepository.CreateAppUserRoles(userRoles.Select(x => (appId, x.Role)).Distinct()),
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/Extensions/AppExtensions.cs:64]: applying Distinct only to the right side of Zip will produce incorrect results. Zip relies on both sides having the same number of elements, both being in the intended order.

File: src/marketplace/Apps.Service/Extensions/AppExtensions.cs [hunk 4]
```diff
@@ -58,11 +58,16 @@ public static void ValidateAppUserRole(Guid appId, IEnumerable<AppUserRole> appU
...
             (AppUserRole appUserRole, UserRole userRole) =>
                 {
-                    userRolesRepository.CreateAppUserRoleDescriptions(appUserRole.Descriptions.Select(appUserRoleDescription => (userRole.Id, appUserRoleDescription.LanguageCode, appUserRoleDescription.Description)));
-                    return new AppRoleData(userRole.Id, appUserRole.Role);
+                    // Since duplicate roles have already been elemenated so, first value of Role's Description amongs the duplicate roles would be selected
+                    var appUserRoleDescriptions = userRoles.First(x => userRole.UserRoleText == x.Role);
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/Extensions/AppExtensions.cs:68]: this attempt to fix the incorrect results of Zip is redundant if Zip is applied correctly.

File: src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs [hunk 5]
```diff
@@ -37,15 +37,33 @@ public UserRolesRepository(PortalDbContext portalDbContext)
 
     public IEnumerable<UserRole> CreateAppUserRoles(IEnumerable<(Guid AppId, string Role)> appIdRoles)
     {
-        var appRoles = appIdRoles.Select(x => new UserRole(Guid.NewGuid(), x.Role, x.AppId)).ToList();
-        _dbContext.AddRange(appRoles);
-        return appRoles;
+        // Checking if given roles already exists against the AppId into the database
+        var appRoles = _dbContext.UserRoles.Where(w => appIdRoles.Select(a => a.AppId).Contains(w.OfferId))
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs:41]: This logic does not belong into the repository, it belongs to the business-logic. Repository should only contain linq-queries that more or less directly translate into sql. The businessLogic would call a suitable query to determine which roles should actually be created, deleted or updated and call the respective repository-methods.

File: src/marketplace/Apps.Service/BusinessLogic/AppReleaseBusinessLogic.cs [hunk 6]
```diff
@@ -170,6 +176,23 @@ public async Task DeleteAppRoleAsync(Guid appId, Guid roleId)
         await _portalRepositories.SaveAsync().ConfigureAwait(ConfigureAwaitOptions.None);
     }
 
+    /// <inheritdoc/>
+    public async Task DeleteAppRolesAsync(Guid appId, IEnumerable<Guid> roleIds)
+    {
+        var companyId = _identityData.CompanyId;
+        var (OfferStatus, IsProviderCompanyUser, RoleIds) = await _portalRepositories.GetInstance<IOfferRepository>().GetAppUserRolesAsync(appId, companyId, OfferStatusId.CREATED, roleIds).ConfigureAwait(ConfigureAwaitOptions.None);
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/BusinessLogic/AppReleaseBusinessLogic.cs:183]: being local variables offerStatus, isProviderCompanyUser, roleIds should be lowercase.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

**Comment by evegufy** [comment]
merge_postponed as a decision about the endpoint semantics need be made, @MaximilianHauer could you please provide your feedback?

File: src/marketplace/Apps.Service/Extensions/AppExtensions.cs (hunk 3 above)
Comment (tfjanjua) [reply src/marketplace/Apps.Service/Extensions/AppExtensions.cs]: Agreed, I tweaked this method a little bit and I tested and it was working fine but the purpose of Zip method got disturbed.
I have reverted this method as it was before.

File: src/marketplace/Apps.Service/Extensions/AppExtensions.cs (hunk 4 above)
Comment (tfjanjua) [reply src/marketplace/Apps.Service/Extensions/AppExtensions.cs]: I have reverted this method as it was before.

File: src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs (hunk 5 above)
Comment (tfjanjua) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/UserRolesRepository.cs]: I have reverted this method as it was before and moved this logic to business-logic.

File: src/marketplace/Apps.Service/BusinessLogic/AppReleaseBusinessLogic.cs (hunk 6 above)
Comment (tfjanjua) [reply src/marketplace/Apps.Service/BusinessLogic/AppReleaseBusinessLogic.cs]: variables has been changed

**Comment by tfjanjua** [comment]
> merge_postponed as a decision about the endpoint semantics need be made, @MaximilianHauer could you please provide your feedback?

Hi @evegufy , do you think we can merge this PR (if all looks good) and then frontend will take their part as per their capacity? because it will not break any frontend functionality but it would be nice if frontend would also work on it and resolve their part (Please check the Proposed Solution heading: https://github.com/eclipse-tractusx/portal-backend/issues/868)
I ask this because we need bug fix from this PR (stopping user from uploading duplicate app user roles) and implementation of new API (to bulk delete recently added App User Roles) is good to have but not urgent.

I can also separate bug (stopping user from uploading duplicate app user roles bug fix) and new API (to bulk delete recently added Roles) in separate PRs so, we could have bug fix quickly available.

Just FYI: we already have frontend capacity available to work on frontend part related to resolving the bug (stopping user from uploading duplicate app user roles bug fix) and implementing the new API (to bulk delete recently added Roles)
https://github.com/eclipse-tractusx/portal-frontend/issues/969

cc: @ybidois | @MaximilianHauer | @ntruchsess | @Phil91

**Comment by evegufy** [comment]
> > merge_postponed as a decision about the endpoint semantics need be made, @MaximilianHauer could you please provide your feedback?
> 
> Hi @evegufy , do you think we can merge this PR (if all looks good) and then frontend will take their part as per their capacity? because it will not break any frontend functionality but it would be nice if frontend would also work on it and resolve their part (Please check the Proposed Solution heading: #868) I ask this because we need bug fix from this PR (stopping user from uploading duplicate app user roles) and implementation of new API (to bulk delete recently added App User Roles) is good to have but not urgent.
> 
> I can also separate bug (stopping user from uploading duplicate app user roles bug fix) and new API (to bulk delete recently added Roles) in separate PRs so, we could have bug fix quickly available.
> 
> Just FYI: we already have frontend capacity available to work on frontend part related to resolving the bug (stopping user from uploading duplicate app user roles bug fix) and implementing the new API (to bulk delete recently added Roles) [eclipse-tractusx/portal-frontend#969](https://github.com/eclipse-tractusx/portal-frontend/issues/969)
> 
> cc: @ybidois | @MaximilianHauer | @ntruchsess | @Phil91

Hi @tfjanjua,

afaik, @ntruchsess @MaximilianHauer where discussing about the implementation approach with the batch delete...
So yes, I think separating the bug and the new API, would help to get the bugfix in sooner.
Before you do that, please wait for feedback from @ntruchsess and @MaximilianHauer.

File: src/marketplace/Apps.Service/Extensions/AppExtensions.cs [hunk 7]
```diff
@@ -56,13 +56,32 @@ public static void ValidateAppUserRole(Guid appId, IEnumerable<AppUserRole> appU
...
+    /// <remarks></remarks>
+    /// <param name="userRolesRepository">repository</param>
+    /// <param name="appId">id of the app</param>
+    /// <param name="userRoles">the app user roles</param>
+    /// <returns>returns the filtered and unique roles</returns>
+    private static async Task<IEnumerable<AppUserRole>> GetUniqueAppUserRoles(IUserRolesRepository userRolesRepository, Guid appId, IEnumerable<AppUserRole> userRoles)
+    {
+        var existingRoles = await userRolesRepository.GetUserRolesForOfferIdAsync(appId).ToListAsync().ConfigureAwait(false);
+        if (existingRoles.Any())
+            userRoles = userRoles.Where(w => !existingRoles.Contains(w.Role));
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/Extensions/AppExtensions.cs:84]: this is reimplementing Linq 'ExceptBy'

File: src/marketplace/Apps.Service/Extensions/AppExtensions.cs [hunk 8]
```diff
@@ -56,13 +56,32 @@ public static void ValidateAppUserRole(Guid appId, IEnumerable<AppUserRole> appU
...
+        return userRoles.GroupBy(g => g.Role).Select(s => s.First());
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/Extensions/AppExtensions.cs:85]: this is not exactly reimplementing 'DistinctBy' but I guess this is what you intent to mimic

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/marketplace/Apps.Service/Extensions/AppExtensions.cs [hunk 9]
```diff
@@ -56,13 +56,32 @@ public static void ValidateAppUserRole(Guid appId, IEnumerable<AppUserRole> appU
     /// <param name="appId">id of the app to create the roles for</param>
     /// <param name="userRoles">the user roles to add</param>
     /// <returns>returns the created appRoleData</returns>
-    public static IEnumerable<AppRoleData> CreateUserRolesWithDescriptions(IUserRolesRepository userRolesRepository, Guid appId, IEnumerable<AppUserRole> userRoles) =>
-        userRoles.Zip(
+    public static async Task<IEnumerable<AppRoleData>> CreateUserRolesWithDescriptions(IUserRolesRepository userRolesRepository, Guid appId, IEnumerable<AppUserRole> userRoles)
```
[the remark this answers is not in the document]
Comment (tfjanjua) [reply src/marketplace/Apps.Service/Extensions/AppExtensions.cs:59]: Thanks for your suggestion @ntruchsess but do you think it make sense to upload Duplicate Roles (like 'Manager' role more than 1 time) with different description? I mean if there are 2 Managers then those should be defined by their proper role name.

**Comment by tfjanjua** [comment]
Hi @evegufy, I discussed with @ntruchsess and he explained the business behind this App User Roles and as per his explanation, it makes sense to replace all the App User Roles when user re-upload them and keep only the new ones instead of all roles (like previously uploaded and recently added from Add Roles option of any Active app) and just ignore the duplicate roles.

So, I have removed the implementation of new API (to delete recently uploaded App User Roles) from this PR and now this PR only has a fix to stop users to upload duplicate roles and we need this fix for our release and later (as we can see @Phil91 has added this PR in release 24.12) we can implement/change the existing API of Upload App User Roles as per @ntruchsess 's explanation.

So, please let me know if this PR can be merge OR should I close this PR and create another one? that new PR will also have 'Stop user to upload duplicate roles' fix.

cc: @ybidois | @MaximilianHauer

**Comment by evegufy** [comment]
> Hi @evegufy, I discussed with @ntruchsess and he explained the business behind this App User Roles and as per his explanation, it makes sense to replace all the App User Roles when user re-upload them and keep only the new ones instead of all roles (like previously uploaded and recently added from Add Roles option of any Active app) and just ignore the duplicate roles.
> 
> So, I have removed the implementation of new API (to delete recently uploaded App User Roles) from this PR and now this PR only has a fix to stop users to upload duplicate roles and we need this fix for our release because if user uploads duplicate roles from any screen (App Onboarding->Technical Integration screen OR Load Additional Roles) so, later ending up encountering 409 conflict errors while assigning those duplicated App User Roles to a Company Users.
> 
> Later (as we can see @Phil91 has added this PR in release 24.12) we can implement/change the existing API of Upload App User Roles as per @ntruchsess 's explanation but this fix will stop users to upload duplicate roles till r24.12.
> 
> So, please let me know if this PR can be merge OR should I close this PR and create another one? that new PR will also have 'Stop user to upload duplicate roles' fix.
> 
> cc: @ybidois | @MaximilianHauer

@tfjanjua good, if it only contains the fix to stop users from uploading duplicate roles, I'll change the milestone to 2.2.0 and the base to the according release branch.
could you please still update the PR description so that it reflects only the fix? Please also update the PR title to a more conventional commit aligned message, like `fix(appRoles): don't allowed to upload duplicate app roles`
@ntruchsess could you please also confirm that this change is fine for you by approving?

**Review by ntruchsess — APPROVED** [review]

**Review by evegufy — APPROVED** [review]

**Pull request merged**